### PR TITLE
Ensure Shibboleth user cannot change their password

### LINKED ISSUE #7959: Azure Active Directory (Shibboleth) user can change their password and login as local DSpace user using their new password
**Description of bug**
Issue is in DSpace7 (unknown if present in previous versions). The issue is on our UAT test instance (we are going live with DSpace7 in Dec 2021 and have not used DSpace before).
We successfully configured authentication with Azure Active Directory (AD) but one requirement of the university is not met, i.e. that AD users must not be able to change their password on DSpace. AD users **can** change their password in DSpace.
AD users can change their password in DSpace user management and later login using local user login page with email and new password. The "new" password is stored in DSpace user management and AD users can bypass AD authentication by logging in with "new" password.

1. Login with AD account
2. Go to profile and change & save new password
3. Logout and close browser
4. Open DSpace again and login as **local** user (local login in our case is visible by adding /login to end of URL) with new password. 
5. Login is successful - which means user bypassed AD authentication, which is not acceptable.

**Expected behavior**
AD users should not be able to change their password in DSpace.

### PR DESCRIPTION
**Pull request by benbosman**
## References
* Fixes https://github.com/DSpace/DSpace/issues/7959

## Description
This add a feature `canChangePassword`, and adds the authentication method to the `authn/status` endpoint

## Instructions for Reviewers
The current authentication method is stored in the JWT token.
This is done by temporarily storing it in the session during the authentication process, and adding it from that session to the token immediately afterwards.

The `canChangePassword` reads this from the JWT token

## Checklist
_This checklist provides a reminder of what we are going to look for when reviewing your PR. You need not complete this checklist prior to creating your PR (draft PRs are always welcome). If you are unsure about an item in the checklist, don't hesitate to ask. We're here to help!_

- [ ] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [ ] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [ ] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [ ] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [ ] If my PR includes new, third-party dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [ ] If my PR modifies the REST API, I've linked to the REST Contract page (or open PR) related to this change.

### CODE REVIEW

**Comment by tdonohue** [comment]
@benbosman : This PR now has minor merge conflicts. My best guess is they are likely caused by the merger of #8032 (or possibly #8031) today.

File: dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/CanChangePasswordFeature.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+@Component
+@AuthorizationFeatureDocumentation(name = CanChangePasswordFeature.NAME,
+        description = "It can be used to verify if the user can change his password")
+public class CanChangePasswordFeature implements AuthorizationFeature {
+
+    public static final String NAME = "canChangePassword";
+
+    @Override
+    public boolean isAuthorized(Context context, BaseObjectRest object) throws SQLException {
+        if (context.getCurrentUser() != null && StringUtils.equals(context.getAuthenticationMethod(), "password")) {
```
Comment (paulo-graca) [inline dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/CanChangePasswordFeature.java:32]: @benbosman does it makes sense to have this configurable? Meaning, have a property to enable/disable this?

Comment (benbosman) [reply dspace-server-webapp/src/main/java/org/dspace/app/rest/authorization/impl/CanChangePasswordFeature.java:32]: @paulo-graca I don't think is was possible in DSpace 6, but making it configurable would only be a very minor change (most of the work goes to ITs to test both options)
I will wait for the Angular work to be completed to make sure there's enough budget left

**Review by tdonohue — CHANGES_REQUESTED** [review]
@benbosman : I gave this a review today & found some minor issues with the code (see inline below). 

However, I also realized that the approach here only fixes one route towards changing your password.  This PR obviously works to block your ability to login via Shibboleth (or a different external system) and change your password from your profile page.  

However, it's still possible for a user who previously authenticated through Shibboleth to click on "Have you forgotten your password?" and then reset/change their password through the forgot password form.  Your approach here of storing your authentication method in the JWT only works if the user is currently authenticated...in the case of a forgot password request, the user would still be able to change their password in DSpace.

The only other approach I can think of is to block password changes if the 'netid' column in the 'eperson' table is not null.  Or in some other way track your *initial* authentication system at the database level.

We should have realized this design flaw earlier...but I only realized it when I went to test this PR and found that I could easily change the Shibboleth user's password using the "Forgot Password" form.

**Comment by tdonohue** [comment]
UPDATE to [my previous review](https://github.com/DSpace/DSpace/pull/8050#pullrequestreview-832034804).

We discussed this in today's DevMtg, and decided that the current approach in this PR is reasonable, and that the proper "fix" for the Forgotten Password issue would be to *disable* the "Have you forgotten your password?" link (or remove it), when ONLY Shibboleth (or LDAP) is enabled, as that'd be similar to 6.x behavior.

So, with regards to my comments above, you can ignore the comments about fixing the Forgotten Password issue.  However, we should still try to resolve the inline comments (missing JavaDocs & improper usage of a HttpSession).

**Review by tdonohue — COMMENTED** [review]
Thanks @YanaDePauw and @benbosman !  I tested this today, and it's working great!

1. Tested using `main` version of UI. Verified that when I try to change my password while logged in via Shibboleth that a 403 error occurs.
2. Tested using UI PR (https://github.com/DSpace/dspace-angular/pull/1454) and verified that the change password options no longer appear when logged in via Shibboleth.

So, overall, I'm a +1.  However, there's *one last* `getSession()` in this PR that was overlooked in the recent updates (see inline below).  Once that is removed as well, I'll submit my formal approval and we can get this moved forward (assuming @paulo-graca also approves)

**Review by paulo-graca — APPROVED** [review]
thank you @benbosman! I did some tests and it works as expected...

using a shib authentication I get no results for `canChangePassword` feature:
`GET /server/api/authz/authorizations/search/object?uri=https://example.com/server/api/eperson/epersons/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73&feature=canChangePassword&embed=feature`
```
{
  "_links": {
    "self": {
      "href": "https://example.com/server/api/authz/authorizations/search/object?uri=https://example.com/server/api/eperson/epersons/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73&feature=canChangePassword"
    }
  },
  "page": {
    "size": 20,
    "totalElements": 0,
    "totalPages": 0,
    "number": 0
  }
}
```

and using a password auth I get 1 result for `canChangePassword` feature:

`GET /server/api/authz/authorizations/search/object?uri=https://example.com/server/api/eperson/epersons/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73&feature=canChangePassword&embed=feature
`
```
{
  "_embedded": {
    "authorizations": [
      {
        "id": "0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73_canChangePassword_eperson.eperson_0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73",
        "type": "authorization",
        "_links": {
          "eperson": {
            "href": "https://example.com/server/api/authz/authorizations/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73_canChangePassword_eperson.eperson_0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73/eperson"
          },
          "feature": {
            "href": "https://example.com/server/api/authz/authorizations/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73_canChangePassword_eperson.eperson_0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73/feature"
          },
          "object": {
            "href": "https://example.com/server/api/authz/authorizations/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73_canChangePassword_eperson.eperson_0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73/object"
          },
          "self": {
            "href": "https://example.com/server/api/authz/authorizations/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73_canChangePassword_eperson.eperson_0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73"
          }
        },
        "_embedded": {
          "feature": {
            "id": "canChangePassword",
            "description": "It can be used to verify if the user can change his password",
            "type": "feature",
            "resourcetypes": [
              "eperson.eperson"
            ],
            "_links": {
              "self": {
                "href": "https://example.com/server/api/authz/features/canChangePassword"
              }
            }
          }
        }
      }
    ]
  },
  "_links": {
    "self": {
      "href": "https://example.com/server/api/authz/authorizations/search/object?uri=https://example.com/server/api/eperson/epersons/0c9bfa63-c1d1-4703-96a8-b6ce39e5fc73&feature=canChangePassword"
    }
  },
  "page": {
    "size": 20,
    "totalElements": 1,
    "totalPages": 1,
    "number": 0
  }
}
```

File: dspace-api/src/main/java/org/dspace/authenticate/X509Authentication.java [hunk 2]
```diff
@@ -563,6 +566,7 @@ public int authenticate(Context context, String username, String password,
                     log.info(LogHelper.getHeader(context, "login",
                                                   "type=x509certificate"));
                     context.setCurrentUser(eperson);
+                    request.getSession().setAttribute(X509_AUTHENTICATED, true);
```
[the remark this answers is not in the document]
Comment (benbosman) [reply dspace-api/src/main/java/org/dspace/authenticate/X509Authentication.java]: This has been adjusted now

**Review by tdonohue — APPROVED** [review]
👍 Looks good to me now! I'll my prior feedback has been addressed and this has been tested by two of us. Thanks @benbosman & @YanaDePauw !

**Pull request merged**